### PR TITLE
DRAFT: Trap processors added

### PR DESCRIPTION
**Pull request by Lehel44**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapReceiver.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    private final ProcessContext context;
+    private final ProcessSession processSession;
+    private final ComponentLog logger;
+
+    public SNMPTrapReceiver(final Snmp snmp, final ProcessContext context, final ProcessSession processSession, final ComponentLog logger) {
+        this.snmp = snmp;
+        this.context = context;
+        this.processSession = processSession;
+        this.logger = logger;
+        snmp.addCommandResponder(this);
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapReceiver.java:48]: "This" should not escape from constructor. Please move this call into some kind of init method.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapReceiver.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public void processPdu(CommandResponderEvent event) {
+        PDU pdu = event.getPDU();
+        if (pdu != null) {
+            FlowFile flowFile = createFlowFile(processSession, pdu);
+            processSession.getProvenanceReporter().receive(flowFile, event.getPeerAddress() + "/" + pdu.getRequestID());
+            if (pdu.getErrorStatus() == PDU.noError) {
+                processSession.transfer(flowFile, REL_SUCCESS);
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapReceiver.java:59]: I am not sure that FF transfer should be anywhere else than a processor. Could you please look after? (same for yield)

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapReceiver.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+    private FlowFile createFlowFile(ProcessSession processSession, PDU pdu) {
+        FlowFile flowFile = processSession.create();
+        flowFile = SNMPUtils.updateFlowFileAttributesWithPduProperties(pdu, flowFile, processSession);
+        return flowFile;
+    }
+
+    public Snmp getSnmp() {
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapReceiver.java]: This method is not used and in general we should not expose internal state if not necessary

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapSender.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        try {
+            PDU pdu;
+            if (target.getVersion() == SnmpConstants.version1) {
+                // Automatically sets type to V1TRAP.
+                PDUv1 pdu1 = new PDUv1();
+                pdu1.setEnterprise(new OID(enterpriseOID));
+                pdu1.setAgentAddress(new IpAddress(agentAddress));
+                pdu1.setGenericTrap(genericTrapType);
+                pdu1.setSpecificTrap(specificTrapType);
+                pdu = pdu1;
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/SNMPTrapSender.java]: pdu1 and pdu2 is pretty unlucky naming. Would not be some more elegant way to do this? (extracting to method?)

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            SNMP_RETRIES,
+            SNMP_TIMEOUT
+    ));
+
+    private static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+            REL_SUCCESS,
+            REL_FAILURE
+    )));
+
+    private SNMPTrapReceiver trapReceiver;
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java]: please use volatile

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        try {
+            snmpClient.listen();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @OnStopped
+    public void stopSnmpClient(ProcessContext context) {
+        trapReceiver = null;
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java:98]: Should not be a `snmp.removeCommandResponder()` called here? (of course within SNMPTrapReceiver in a close method)

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) {
+        if (Objects.isNull(trapReceiver)) {
+            trapReceiver = new SNMPTrapReceiver(snmpClient, context, session, getLogger());
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java]: I am not sure how this is gonna react on processor settings like how frequently it should be triggered or if it should be triggered on primary node only. Sending FFs directly in the receiver will override all of these aspects. I think the preferred way would be:
1. The receiver would collect the incoming events into a buffer (BlockingQueue?)
2. When triggered, the processor should pick up them and transfer them.
Of course this comes with managing empty and full buffer, etc.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    @OnStopped
+    public void stopSnmpClient(ProcessContext context) {
+        trapReceiver = null;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) {
+        if (Objects.isNull(trapReceiver)) {
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java:103]: Is there a chance that this is null after onScheduled?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                .setPrivacyPassword(context.getProperty(SNMP_PRIVACY_PASSWORD).getValue())
+                .setSecurityName(context.getProperty(SNMP_SECURITY_NAME).getValue())
+                .setSecurityLevel(context.getProperty(SNMP_SECURITY_LEVEL).getValue())
+                .createSecurityConfiguration();
+
+        snmpClient = SNMPClientFactory.createSnmpClient(configuration, clientPort);
+        try {
+            snmpClient.listen();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java]: I think it should throw a ProcessException (to stop scheduling)

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+public class ListenTrapSNMP extends AbstractSNMPProcessor {
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java:54]: A number of annotation should be put here. Annotations like: InputRequirements, Tags, Description, SupportsBatching and Restricted

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+import org.snmp4j.smi.OID;
+import org.snmp4j.smi.TimeTicks;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class SendTrapSNMP extends AbstractSNMPProcessor {
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java:50]: Please add annotations as well

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor GENERIC_TRAP_TYPE = new PropertyDescriptor.Builder()
+            .name("snmp-trap-generic-type")
+            .displayName("Generic trap type")
+            .description("Generic trap type is an integer in the range of 0 to 6. See Usage for details.")
+            .required(true)
+            .allowableValues("0", "1", "2", "3", "4", "5", "6")
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java:73]: `StandardValidators.createLongValidator(0, 6, true)`

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+            .build();
+
+    public static final PropertyDescriptor SPECIFIC_TRAP_TYPE = new PropertyDescriptor.Builder()
+            .name("snmp-trap-specific-type")
+            .displayName("Specific trap type")
+            .description("Specific trap type is a number that further specifies the nature of the event that generated " +
+                    "the trap in the case of traps of generic type 6 (enterpriseSpecific). The interpretation of this " +
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java:81]: dependOn()

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor MANAGER_ADDRESS = new PropertyDescriptor.Builder()
+            .name("snmp-trap-manager-address")
+            .displayName("Manager (destination) IP address")
+            .description("The address of the SNMP manager where the trap is being sent to.")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java:109]: Is not there any more specific validator for IPs?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+            .build();
+
+    public static final PropertyDescriptor SYSTEM_UPTIME = new PropertyDescriptor.Builder()
+            .name("snmp-trap-sysUpTime")
+            .displayName("System uptime")
+            .description("The system uptime.")
+            .required(true)
+            .addValidator(StandardValidators.INTEGER_VALIDATOR)
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java:117]: StandardValidators.POSITIVE_INTEGER_VALIDATOR

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+     * {@link #onTrigger(ProcessContext, ProcessSession)}. It is implemented by
+     * sub-classes to perform {@link Processor} specific functionality.
+     *
+     * @param context        instance of {@link ProcessContext}
+     * @param processSession instance of {@link ProcessSession}
+     * @throws ProcessException Process exception
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession processSession) {
+        final String enterpriseOIDValue = context.getProperty(ENTERPRISE_OID).getValue();
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java]: Would not make sense to extract these properties during onSchedule and depend on those later?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+            TRAP_OID_VALUE,
+            SYSTEM_UPTIME
+    ));
+
+    private static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+            REL_SUCCESS,
+            REL_FAILURE
+    )));
+
+    private SNMPTrapSender snmpTrapSender;
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SendTrapSNMP.java]: Please use volatile

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/ListenTrapSNMPTest.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        String managerAddress = "1.2.3.5";
+        String trapOIDValue = "TrapOidValue";
+        TimeTicks sysUpTime = new TimeTicks(5000);
+
+        Snmp snmp = SNMPClientFactory.createSnmpClient(String.valueOf(NetworkUtils.availablePort()));
+
+        try {
+            SNMPTrapSender trapSender = new SNMPTrapSender(snmp, target);
+            trapSender.generateTrap(sysUpTime, enterpriseOID, agentAddress, genericTrapType, specificTrapType, trapOID, managerAddress, trapOIDValue);
+            Thread.sleep(200);
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/ListenTrapSNMPTest.java]: Why this is necessary?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/SendTrapSNMPTest.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class SendTrapSNMPTest {
+
+    @Test
+    public void testReceive() throws IOException, InterruptedException {
+
+        int trapSenderClientPort = NetworkUtils.availablePort();
```
Comment (simonbence) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/SendTrapSNMPTest.java]: I would suggest to create an abstract parent class for these tests to reduce duplication

**Review by simonbence — CHANGES_REQUESTED** [review]
There are some questions, could you please look after them?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java (hunk 6 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java:98]: Doublecheck

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java (hunk 8 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java:103]: Yes, there's no @OnScheduled call. I must have initialized this here to pass the context and session.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java (hunk 6 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/ListenTrapSNMP.java:98]: When the processor is started, the @OnScheduled init will create a new snmp instance every time. I think it's not needed.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/ListenTrapSNMPTest.java (hunk 18 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/ListenTrapSNMPTest.java]: SNMP requests are asynchronous. I checked and 20 ms is enough for the message to arrive. I found awaitility complicated at first glance and this is the only place where a little stop is needed.

**Pull request closed without merging**